Sources/SwiftFusion/Core/TrappingDouble.swift

```diff
@@ -0,0 +1,232 @@
+// Copyright 2020 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import PythonKit
+import TensorFlow
+
+/// A wrapper for `Double` that traps instead of allowing `NaN`.
+public struct TrappingDouble: AdditiveArithmetic, Differentiable, Equatable {
+  public var value: Swift.Double
+
+  public init(_ value: Swift.Double) {
+    assert(!value.isNaN)
+    self.value = value
+  }
+}
+
+/// A wrapper for `Double` that traps instead of allowing `NaN`.
+///
+/// This typealias makes it easier to use unmodified existing code with `TrappingDouble`.
+///
+/// If you really need the normal Swift `Double`, use `Swift.Double` to circumvent this typealias.
+public typealias Double = TrappingDouble
```

Comment:
I think we should have a RelWithAssertion build before globally enabling this?

Reply:
Not sure what you mean. If we merge this now, then it'll trap NaNs on debug builds but not on release builds, which seems fine.

I do think we should comment out or `#if` out this line before merging, so that we don't have a weird nonstandard Double all the time. People debugging NaNs can enable the typealias locally.

Reply:
Yup, my bad for the phrasing :) Let's `#if` out this before merging.